### PR TITLE
Expose AnalysisContext.MinimumReportedSeverity

### PR DESCRIPTION
**Pull request by mavasani**
Work towards #52991

Implements the compiler/analyzer driver performance change from https://github.com/dotnet/roslyn/issues/52991#issuecomment-1709912198 This flag indicates the minimum reported diagnostic severity for analyzer execution context. For command line builds, this will be warning, while for the IDE case this will be hidden. Analyzer diagnostics with severity lesser than this severity are not reported by the driver.

This flag will allow the IDE code style analyzers to turn on/off in command line builds based on `option_name = option_value:severity` entries in editorconfig.

### CODE REVIEW

**Comment by 333fred** [comment]
@mavasani is there a public api proposal for this change?

**Comment by mavasani** [comment]
> @mavasani is there a public api proposal for this change?

Ah, this did not go through the public API review process, but was approved few months back at a one-off design meeting including members from the compiler and IDE team for this feature. Let me add an API proposal to the issue and mark it ready for api-review.

**Comment by mavasani** [comment]
Moving to draft until we have have an API approval

**Comment by mavasani** [comment]
> @mavasani is there a public api proposal for this change?

@333fred I have added a public API proposal [here](https://github.com/dotnet/roslyn/issues/52991#issuecomment-1742711060) and marked the issue with `api-ready-for-review`. Thanks!

**Comment by mavasani** [comment]
> > @mavasani is there a public api proposal for this change?
> 
> @333fred I have added a public API proposal [here](https://github.com/dotnet/roslyn/issues/52991#issuecomment-1742711060) and marked the issue with `api-ready-for-review`. Thanks!

The public API was approved in https://github.com/dotnet/roslyn/issues/52991#issuecomment-1760359293, marking as ready for review.

**Review by 333fred — APPROVED** [review]

**Comment by mavasani** [comment]
Thank you for the reviews!

**Pull request merged**
